src/redis.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Define the Superset server Postgresql relation."""
+
+import logging
+
+from charms.redis_k8s.v0.redis import RedisRelationCharmEvents
+from ops import framework
+from ops.framework import StoredState
+
+from log import log_event_handler
+
+logger = logging.getLogger(__name__)
+
+
+class Redis(framework.Object):
+    """Client for superset:redis relation.
+
+    Attrs:
+        on: redis relation events from redis_k8s library
+        _stored: charm stored state
+    """
+
+    on = RedisRelationCharmEvents()
+    _stored = StoredState()
+
+    def __init__(self, charm):
+        """Construct.
+
+        Args:
+            charm: The charm to attach the hooks to.
+        """
+        super().__init__(charm, "redis")
+        self.charm = charm
+        self.framework.observe(
+            charm.on.redis_relation_updated, self._on_redis_relation_changed
+        )
+
+    @log_event_handler(logger)
+    def _on_redis_relation_changed(self, event):
+        """Handle redis relation updated event.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if self.charm._stored.redis_relation:
+            host, port = self._get_redis_relation_data()
+            self.charm._state.redis_relation = "enabled"
+        else:
+            host = None
+            port = None
+            self.charm._state.redis_relation = "disabled"
```

Comment:
**suggestion:** Consider using some `relation` `property` here that returns a bool on the relation.

It's down to opinion, but to me, this:
```python
# charm.py example
if not self.redis.relation:
   do_stuff()
```

Reply:
My understanding (let me know if I have misunderstood!) is there are 2 points here: 
1) removing disabled/enabled in favour of a bool value.
2) removing the reliance on `self._state` in favour of a relation specific property `self.redis`. So instead of what I currently have I will end up with: `self.redis.relation`, `self.redis.host` and `self.redis.port`

For 1, I agree and will make the update. For 2, we use `_state` in this way consistently across a number of our charms at the moment, I'm not entirely clear on the pros and cons of using this vs a relation specific property. I believe `_state` uses the peer relation as a data store which enables us to trigger methods with `on_peer_relation_changed`. 

@kelkawi-a for visibility as this conversation is relevant for his charms also.

Reply:
Yeah that's a good summary, you've got it. 

IMO, having a completely un-typed, dynamic object across everything in `CharmBase` rapidly gets difficult to remember. If people leave the project, all that 'knowledge' of the various key-names that persist there leaves with them, and makes it much harder to maintain and support.

If I understand correctly, this isn't a mutable attr that NEEDS to be shared, it functions as a convenience collection of data. If so, it makes much more sense to 'route' through `CharmBase`:


```
# REDIS                       # TLS
self.host                     self.cert
self.charm.tls.cert           self.charm.redis.host
			       \          /
                     # charm
                     self.redis.host
                     self.tls.cert 
```

Reply:
If you want to keep the convenience of setting this stuff, you can always add a [`setter`](https://github.com/canonical/data-platform-libs/blob/2f76b5a283ce7a6b084bab515b9d248f57a9450c/lib/charms/data_platform_libs/v0/upgrade.py#L768-L779)

Reply:
Ok I think I've understood. So you're suggesting instead of updating `self.charm._state` and accessing this across everything. Instead within our relations we update `self.{attribute}` and then this can be accessed in `CharmBase` using `self.{relation}.{attribute}` or in another relation using `self.charm.{relation}.{attribute}`. 

The main difference seems to be that we are using the peer relation "databag" for everything instead of the relation specific databags. Because that is all `_state` is doing: https://github.com/canonical/superset-k8s-operator/blob/main/src/state.py#L28. I'm not 100% on why the peer relation data store would be untyped and dynamic if the relation specific one is not? Seems to me they should operate in a similar way.

Reply:
As discussed, this topic is bigger than this specific PR so sticking with `_state` for now and will discuss in more detail next week.